src/D2L.CodeStyle.Analyzers/Immutability/ImmutableTypeKind.cs

```diff
@@ -0,0 +1,28 @@
+﻿using System;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	/// <summary>
+	/// Our "immutability" is a property of values. However, we can also
+	/// generalize this to types as a whole to varying degrees.
+	/// </summary>
+	[Flags]
+	internal enum ImmutableTypeKind {
+		/// <summary>
+		/// We have nothing to say about the immutability of this type.
+		/// </summary>
+		None = 0,
+
+		/// <summary>
+		/// A type T has immutable instances if "new T( ... )" always produces
+		/// an immutable value.
+		/// </summary>
+		Instance = 1,
+
+
+		/// <summary>
+		/// A type T is totally immutable if all objects of type T are
+		/// immutable. This is a stronger condition than instance immutability.
+		/// </summary>
+		Total = 2 | Instance,

```

Comment:
This+`ImmutableTypeInfo` is the stuff I care most about pinning down: we have a shocking amount of reimplementation in the immutability stuff due to how it was developed over time. Stuff like `[ImmutableGeneric]` and the many ways that we "bless" types.

The comments in here I think capture what we're trying to do. To make all this happen we need:

* A way to tell if a type is known to be immutable in one of these two ways (`ImmutabilityContext` -- it is context dependent, e.g. based on type arguments, and when you're analyzing `Foo`s decl you assume that `Foo` is immutable, etc.)
* A way to validate that an `[Immutable]` decl is valid (looks at fields, makes sure they're readonly, etc.... `ImmutableDeclarationChecker` in next PR)
* If a type is `[Immutable]` (in this `Total` sense) then any subclass/implementation also needs to be `[Immutable]`. We already have an analyzer for this (it was done awhile ago with this refactoring in mind)